src/theory/quantifiers/sygus/sygus_random_enumerator.cpp

```diff
@@ -0,0 +1,199 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Abdalrhman Mohamed, Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Implementation of sygus random enumerator.
+ */
+
+#include "theory/quantifiers/sygus/sygus_random_enumerator.h"
+
+#include "expr/dtype_cons.h"
+#include "expr/skolem_manager.h"
+#include "options/quantifiers_options.h"
+#include "theory/datatypes/sygus_datatype_utils.h"
+#include "theory/quantifiers/sygus/term_database_sygus.h"
+#include "theory/rewriter.h"
+#include "util/random.h"
+
+namespace cvc5 {
+namespace theory {
+namespace quantifiers {
+
+void SygusRandomEnumerator::initialize(Node e)
+{
+  d_tn = e.getType();
+  Assert(d_tn.isDatatype());
+  Assert(d_tn.getDType().isSygus());
+  SygusTypeInfo sti;
+  sti.initialize(d_tds, d_tn);
+  std::vector<TypeNode> stns;
+  sti.getSubfieldTypes(stns);
+  // We will be using the datatype constructors a lot, so we cache them here.
+  for (const TypeNode& stn : stns)
+  {
+    for (const std::shared_ptr<DTypeConstructor>& cons :
+         stn.getDType().getConstructors())
+    {
+      if (cons->getNumArgs() == 0)
+      {
+        d_noArgCons[stn].push_back(cons);
+      }
+      else
+      {
+        d_argCons[stn].push_back(cons);
+      }
+    }
+  }
+}
+
+bool SygusRandomEnumerator::increment()
+{
+  Node n, bn;
+  do
+  {
+    // Generate the next sygus term.
+    n = incrementH();
+    bn = d_tds->sygusToBuiltin(n);
+    bn = Rewriter::callExtendedRewrite(bn);
+    // Ensure that the builtin counterpart is unique (up to rewriting).
+  } while (d_cache.find(bn) != d_cache.cend());
+  d_cache.insert(bn);
+  d_currTerm = n;
+  return true;
+}
+
+Node SygusRandomEnumerator::incrementH()
+{
+  NodeManager* nm = NodeManager::currentNM();
+  SkolemManager* sm = NodeManager::currentNM()->getSkolemManager();
+  Random& rand = Random::getRandom();
```

Comment:
This is a pretty confusing name, `rand()` is builtin C++ method, you are also using `rand()` below, which is not clear which is getting called.

Reply:
I named it `rand` at first because I thought no one cares about implementation details. Though, it does bring some maintenance issues now that I think about it. Changed to `rnd`.